src/rt/GEAR/rt_grackle_cooling.h

```diff
@@ -0,0 +1,228 @@
+ * @brief header file for the GEAR M1 closure radiative transfer scheme
+ * kiara grackle cooling related functions.
+ */
+
+/**
+ * @brief Returns the value of G0 for given particle p
+ *
+ * @param p Pointer to the particle data.
+ * @param cooling The properties of the cooling function.
+ *
+ */
+__attribute__((always_inline)) INLINE float cooling_compute_G0(
+                const struct part *restrict p,
+                const struct rt_props* rt_props) {
+
+      float G0 = 0.f;
+      /* Determine ISRF in Habing units based on chosen method */
+      if (rt_props->G0_computation_method==0) {
+          G0 = 0.f;
+      }
+      else if (rt_props->G0_computation_method==1) {
+          G0 = p->chemistry_data.local_sfr_density * rt_props->G0_factor1;
+      }
+      else if (rt_props->G0_computation_method==2) {
+          G0 = p->group_data.ssfr * rt_props->G0_factor2;
+      }
+      else if (rt_props->G0_computation_method==3) {
+          if (p->group_data.ssfr > 0.) {
+              G0 = p->group_data.ssfr * rt_props->G0_factor2;
+          }
+          else {
+              G0 = p->chemistry_data.local_sfr_density * rt_props->G0_factor1;
+          }
+      }
+      return G0;
+}
+
+/**
+ * @brief get the densities of all species and electrons.
+ *
+ * @param p particle to use
+ * @param rho particle physical density
+ * @param species_densities array to write densities in
+ *
+ **/
+__attribute__((always_inline)) INLINE static void
+rt_tchem_get_species_densities(const struct part* restrict p, gr_float rho,
+                               gr_float species_densities[RT_N_SPECIES],
+                               gr_float species_extra[rt_species_extra_count],
+                               const struct rt_props* rt_props) {
+    
+    /* nHII = rho_HII / m_p
+     * nHeII = rho_HeII / 4 m_p
+     * nHeIII = rho_HeIII / 4 m_p
+     * ne = nHII + nHeII + 2 * nHeIII
+     * But: it is grackle convention to use rho_e = n_e * m_p */
+  
+  for (enum rt_species species = 0; species < rt_species_count; species++){
+        species_densities[species] = p->rt_data.tchem.mass_fraction[species] * rho;
+  }
+
+#ifdef SWIFT_RT_GRACKLE_DUST
+  /* Load dust and metal info */ 
+  species_extra[rt_species_dust] = p->rt_data.cooling.dust_mass / p->mass * rho;
+  species_extra[rt_species_SNe_ThisTimeStep] = p->feedback_data.SNe_ThisTimeStep;
+  //if( chemistry_get_total_metal_mass_fraction_for_cooling(p)>0.f) message("Zsm= %g Zp= %g Z= %g Zd= %g",chemistry_get_total_metal_mass_fraction_for_cooling(p), p->chemistry_data.metal_mass_fraction_total, species_densities[19], species_densities[20]);
+  
+  /* Determine ISRF in Habing units based on chosen method */
+  species_extra[rt_species_isrf_habing] = cooling_compute_G0(p, rt_props);
+  
+  /* Load gas metallicities NEED TO CHANGE TO RIGHT COUNT  */
+  for (enum rt_species_extra species = rt_species_He_gas; species < rt_species_He_dust; species++){
+        int i = 0;
```

Comment:
This is wrong. For each iteration of your for loop, `i` will always be zero. You increment it at the end of your loop, but it will be reset at the start of the next iteration.

Reply:
Yes, I place it outside the loop now